### PR TITLE
Cache empty arrays

### PR DESCRIPTION
**Pull request by Maxwe11**
Caching non empty arrays in progress :smile:

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs [hunk 1]
```diff
@@ -76,7 +76,7 @@ private void HandleTypeDeclaration(SyntaxNodeAnalysisContext context)
             {
                 if (!XmlCommentHelper.HasDocumentation(declaration))
                 {
-                    context.ReportDiagnostic(Diagnostic.Create(Descriptor, declaration.Identifier.GetLocation()));
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptor, declaration.Identifier.GetLocation(), ArrayEx.Empty<object>()));
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: :exclamation: This is handled by Roslyn in cases where mscorlib contains `Array.Empty<T>`.  I don't want to make the code more complex to address this. If we need it, we can target .NET 4.6 or another framework which contains it.

**Comment by sharwell** [comment]
I didn't see any instances other than the fixes to `params` arguments. As such, based on my previous comment I'm closing as **wontfix**.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs (hunk 1 above)
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: Any links with info about this?

Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: * dotnet/roslyn#1103
* dotnet/roslyn#1117

Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: But I don't know what ver of mscorlib used there
http://tryroslyn.azurewebsites.net/#K4Zwlgdg5gBAygTxAFwKYFsDcAoADsAIwBswBjGUogQxBBgGEYBvbGNmfYsmANwHswAExgARABQBKZq3ayAspJyyAvjLZqOhEuX5CYC3FQBOVdHT4EAVqlLIA2gF0YfKSxXZlQAA

Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: It's one that doesn't have `Array.Empty<T>()` (it doesn't compile even if you try to reference it directly).

Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1600ElementsMustBeDocumented.cs]: :memo: Also note that all of these cases only occur in code that has violations. Clean code, which is the most important case to optimize for, does not benefit from these changes at all.

**Pull request closed without merging**